File: c++/384. Shuffle an Array.cpp
class Solution {
public:
    Solution(vector<int> nums) : nums(nums) {
        
    }
    
    /** Resets the array to its original configuration and return it. */
    vector<int> reset() {
        return nums;
    }
    
    /** Returns a random shuffling of the array. */
    vector<int> shuffle() {
        vector<int> res(nums);
        int n = res.size();
        for (int i = 0; i < n; i++) {
            int j = rand()%(n-i);
            swap(res[i+j], res[i]);
        }
        return res;
    }

private:
    vector<int> nums;
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(nums);
 * vector<int> param_1 = obj.reset();
 * vector<int> param_2 = obj.shuffle();
 */

File: go/src/leetcode/384. Shuffle an Array/384. Shuffle an Array.go
/**
@author ZhengHao Lou
Date    2021/11/22
*/
package main

import "math/rand"

/**
https://leetcode-cn.com/problems/shuffle-an-array/
 */
type Solution struct {
	n int
	nums []int
}


func Constructor(nums []int) Solution {
	return Solution{
		n: len(nums),
		nums: nums,
	}
}


func (this *Solution) Reset() []int {
	return this.nums
}


func (this *Solution) Shuffle() []int {
	shuffle := append([]int{}, this.nums...)
	for i := range this.nums {
		j := rand.Intn(this.n - i)
		shuffle[i], shuffle[i+j] = shuffle[i+j], shuffle[i]
	}
	return shuffle
}


/**
 * Your Solution object will be instantiated and called as such:
 * obj := Constructor(nums);
 * param_1 := obj.Reset();
 * param_2 := obj.Shuffle();
 */
